src/zino/tasks/bgpstatemonitortask.py

```diff
@@ -0,0 +1,420 @@
+
+from zino.oid import OID
+from zino.snmp import SNMP, MibObject, NoSuchNameError, SparseWalkResponse
+from zino.statemodels import BgpAdminStatus, BGPEvent, BgpOperState, BgpStyle, IPAddress
+from zino.tasks.task import Task
+
+_logger = logging.getLogger(__name__)
+
+TIME_BEFORE_OPER_DOWN_ALERT = 600
+
+
+@dataclass
+class BaseBgpRow:
+    peer_state: BgpOperState
+    peer_admin_status: BgpAdminStatus
+    peer_remote_address: IPAddress
+    peer_remote_as: int
+    peer_fsm_established_time: int
+
+
+class BgpStateMonitorTask(Task):
+    """Fetches and stores state information about external BGP sessions."""
+
+    async def run(self):
+        bgp_style = await self._get_bgp_style()
+        if bgp_style != self.device_state.bgp_style:
+            _logger.debug(f"Router {self.device_state.name} wrong BGP style, was '{self.device_state.bgp_style}'")
+            self.device_state.bgp_style = bgp_style
+        if not bgp_style:
+            return
+
+        local_as = await self._get_local_as(bgp_style=bgp_style)
+        uptime = await self._get_uptime()
+        if not local_as or not uptime:
+            return
+
+        if bgp_style == "juniper":
+            bgp_info = await self._get_juniper_bgp_info()
+        elif bgp_style == "cisco":
+            bgp_info = await self._get_cisco_bgp_info()
+        elif bgp_style == "general":
+            bgp_info = await self._get_general_bgp_info()
+
+        if not bgp_info:
+            return
+
+        bgp_info = self._fixup_ip_addresses(bgp_info=bgp_info)
+
+        for oid, result in bgp_info.items():
+            self._update_single_bgp_entry(oid=oid, row=result, local_as=local_as, uptime=uptime)
+
+    async def _get_bgp_style(self) -> Union[BgpStyle, None]:
+        snmp = SNMP(self.device)
+
+        # TODO Should this be moved to the SNMP file?
+        def _is_in_scope(entry: MibObject, oid: Tuple[str, str]):
+            object_type = snmp._oid_to_object_type(*oid)
+            snmp._resolve_object(object_type=object_type)
+            root = OID(object_type[0])
+            return root.is_a_prefix_of(other=entry.oid)
+
+        try:
+            entry = await snmp.getnext("BGP4-V2-MIB-JUNIPER", "jnxBgpM2")
+        except NoSuchNameError:
+            pass
+        else:
+            if entry and _is_in_scope(entry=entry, oid=("BGP4-V2-MIB-JUNIPER", "jnxBgpM2")):
+                return "juniper"
+
+        try:
+            entry = await snmp.getnext("CISCO-BGP4-MIB", "cbgpPeer2Table")
+        except NoSuchNameError:
+            pass
+        else:
+            if entry and _is_in_scope(entry=entry, oid=("CISCO-BGP4-MIB", "cbgpPeer2Table")):
+                return "cisco"
+
+        try:
+            entry = await snmp.getnext("BGP4-MIB", "bgp")
+        except NoSuchNameError:
+            pass
+        else:
+            if entry and _is_in_scope(entry=entry, oid=("BGP4-MIB", "bgp")):
+                return "general"
+
+        return None
+
```

Comment:
Following the rule of "three times is a pattern", this should be boiled down to something simpler, and the suggestion for making a generic test-for-support method would help in that regard